### PR TITLE
Between thresholds plugin

### PR DESCRIPTION
**Pull request by cgsandford**
Plugin to calculate a "between_thresholds" cube from above or below thresholded probabilities.

The question of whether threshold points should be at the top, bottom or midrange of the bounds has been asked but not answered yet.  In the meantime I've written the plugin to have the upper threshold point and bounds spanning the range between which probabilities are specified.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/tests/between_thresholds/test_between_thresholds.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    def test_error_thresholds_unavailable(self):
+        """Test error if cube doesn't contain the required thresholds"""
+        threshold_ranges = [[10, 100], [1000, 30000]]
+        plugin = OccurrenceBetweenThresholds(threshold_ranges, 'm')
+        msg = ('visibility threshold 10 is not available\n'
+               'visibility threshold 30000 is not available')
+        with self.assertRaisesRegex(ValueError, msg):
+            plugin.process(self.vis_cube)
+
```
Comment (bayliffe) [inline lib/improver/tests/between_thresholds/test_between_thresholds.py]: We need a test with a zero threshold to capture the difficulties of using np.close.

File: lib/improver/between_thresholds.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        """
+        multiplier = self._get_multiplier(thresh_coord)
+        thresh_name = thresh_coord.name()
+
+        cubelist = iris.cube.CubeList([])
+        for (lower_cube, upper_cube) in self.cube_slices:
+            # construct difference cube
+            between_thresholds_data = (
+                lower_cube.data-upper_cube.data)*multiplier
+            between_thresholds_cube = upper_cube.copy(between_thresholds_data)
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/between_thresholds.py]: I have asked Bruce the question about how to label these points, but I think he's on leave this week.  We can certainly address this as part of that discussion.  IMO the main purpose of this plugin will be visibility risk bands - but if we decide to label by "top" we should definitely document what the plugin will not support.

File: lib/improver/between_thresholds.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        cubes = []
+        for t_range in self.threshold_ranges:
+            t_range.sort()
+            lower_constraint = iris.Constraint(coord_values={
+                thresh_coord: lambda t: np.isclose(t.point, t_range[0])})
+            lower_cube = cube.extract(lower_constraint)
+            if lower_cube is None:
+                error_msg += error_string.format(t_range[0])
+            upper_constraint = iris.Constraint(coord_values={
+                thresh_coord: lambda t: np.isclose(t.point, t_range[1])})
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/between_thresholds.py]: Good point.  Will specify a sensible `atol` to address this.

File: lib/improver/between_thresholds.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            cube (iris.cube.Cube):
+                Probability cube containing thresholded data (above or below)
+
+        Returns:
+            output_cube (iris.cube.Cube):
+                Cube containing probability of occurrence between thresholds
+        """
+        # if cube has no threshold-type coordinate, raise an error
+        try:
+            thresh_coord = find_threshold_coordinate(cube)
```
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/between_thresholds.py]: I need to interrogate the coordinate attributes for `_get_multiplier`, which is why I have set this variable as the whole coordinate rather than the name.  However it does make sense to make this a class attribute.  I'll update this.

File: lib/improver/between_thresholds.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/between_thresholds.py]: OK, actually I had no issue with this - the unit test I wrote for this passed straight away.  But if it doesn't work on all hardware, I'll try setting `atol` to `10^-6` for this case.

File: lib/improver/tests/between_thresholds/test_between_thresholds.py (hunk 1 above)
Comment (cgsandford) [reply lib/improver/tests/between_thresholds/test_between_thresholds.py]: OK

File: lib/improver/between_thresholds.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/between_thresholds.py]: Ah actually I'm not going to unless I need to, for the reason that if we use this for precipitation rates in `m s-1`, this will be a problem.

[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/between_thresholds.py]: Your test passes because you have not made the coordinate in your unit test of type float32. If you change it as below you will find that the test fails and you still need to handle this.
```
self.temp_cube.coord('air_temperature').points = np.array([
    272.15, 273.15, 274.15, 275.15], dtype=np.float32)
```
Good old float32.

**Comment by bayliffe** [comment]
@cgsandford 
Largely happy, but I think you still need to deal with 0 thresholds (see comment regarding your test).

File: lib/improver/between_thresholds.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/between_thresholds.py]: Eurgh.  This requires an absolute tolerance of `1e-5`.  I'll make a note in the docstring.

[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/between_thresholds.py]: It's not lovely. I'm struggling to think of much else we can do without making horrible special cases. The horrible special cases treatment might be something like adding 1 to the desired threshold and coordinate when the abs(threshold) < 1.0E-4. This would allow rtol to play a role in slightly expanding the tolerance without having to blanket raise atol, but this is a horrible fudge fix.

End of my musings. I think what you've done will have to suffice for now.

File: lib/improver/between_thresholds.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        # update output cube metadata
+        output_cube.rename(
+            'probability_of_{}_between_thresholds'.format(
+                extract_diagnostic_name(cube.name())))
+        new_thresh_coord = output_cube.coord(self.thresh_coord.name())
+        new_thresh_coord.convert_units(original_units)
+        new_thresh_coord.attributes['spp__relative_to_threshold'] = (
+            'between_thresholds')
+
+        return output_cube
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply lib/improver/between_thresholds.py]: In discussion, we've decided to ignore this.

File: lib/improver/tests/between_thresholds/test_between_thresholds.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+import iris
+from iris.tests import IrisTest
+
+from improver.tests.set_up_test_cubes import (
+    set_up_probability_cube, set_up_percentile_cube)
+from improver.between_thresholds import OccurrenceBetweenThresholds
+
+
+class Test_process(IrisTest):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply lib/improver/tests/between_thresholds/test_between_thresholds.py]: @MoseleyS @cgsandford @gavinevans 
Why were the tests not split up by function in this case?

[the remark this answers is not in the document]
Comment (cgsandford) [reply lib/improver/tests/between_thresholds/test_between_thresholds.py]: @bayliffe The conceptual approach to unit tests is to test all paths through the code that are publically accessible.  The only public interface in the class is `process`.  The functions are factored in the way they are to make `process` clear, and all paths through the sub-methods are tested in `Test_process`.

In many of our IMPROVER classes we repeat tests at multiple levels, essentially having the same path tested both at the `process` and private method level, leading to a huge amount of code to maintain, which also has to be radically changed if we refactor the internal workings of a class, even if the public interface and behaviour remains identical.  The way I have written these tests means that if someone decides there is a better way to factor this internally and they don't change the class behaviour, they shouldn't have to touch the unit tests at all.  IMO this makes the code base both cleaner (in terms of focusing on the important behaviour of the public interface) and more maintainable (refactoring is less time-intensive).

Although this is different from the way we have traditionally done it, this approach is within the bounds defined here https://github.com/metoppv/improver/wiki/Code-Style-Guide#unit-testing.  We also discussed it with @benfitzpatrick to confirm this.
